### PR TITLE
Custom compile time quadrature rules

### PR DESCRIPTION
**Pull request by michalhabera**
Allows a user to pass his own set of points and weights as metadata in integral.

The syntax is following
```python
points = [[0.0, 0.0], [1.0, 0.0], [0.0, 1.0], [0.5, 0.5], [0.0, 0.5], [0.5, 0.0]]
weights = [1 / 12] * 6
u * v * ufl.dx(metadata={"quadrature_rule": "custom", "quadrature_points": points, "quadrature_weights": weights})
```

### CODE REVIEW

File: ffcx/ir/representation.py [hunk 1]
```diff
@@ -422,8 +423,37 @@ def _compute_integral_ir(form_data, form_index, prefix, element_numbers, integra
             scheme = md["quadrature_rule"]
             degree = md["quadrature_degree"]
 
-            (points, weights) = create_quadrature_points_and_weights(integral_type, cell, degree,
-                                                                     scheme)
+            if scheme == "custom":
+                points = md["quadrature_points"]
+                weights = md["quadrature_weights"]
+            elif scheme == "vertex":
+                # FIXME: Could this come from FIAT?
```
Comment (miklos1) [inline ffcx/ir/representation.py:430]: Quite certainly yes, and it should.

File: ffcx/ir/representation.py [hunk 2]
```diff
@@ -422,8 +423,37 @@ def _compute_integral_ir(form_data, form_index, prefix, element_numbers, integra
             scheme = md["quadrature_rule"]
             degree = md["quadrature_degree"]
 
-            (points, weights) = create_quadrature_points_and_weights(integral_type, cell, degree,
-                                                                     scheme)
+            if scheme == "custom":
+                points = md["quadrature_points"]
+                weights = md["quadrature_weights"]
```
Comment (chrisrichardson) [inline ffcx/ir/representation.py:428]: Should there be some checks here, (1) that the points and weights have been provided (and if not, return a useful message), (2) that they are the same length?

**Comment by mrambausek** [comment]
As I am heavily relying on QuadratureElement, are there any plans/ideas of extending them to custom quadrature rules?

**Comment by michalhabera** [comment]
@mrambausek We need to review "QuadratureElement" codepath again. I can have a look at custom points in that case as well.

**Pull request merged**